File: frontend/src/main/offline/antrean.js
'use strict'

// Antrean kirim (offline fase 2, padanan Android `core/offline/antrean.dart`):
// permintaan tulis yang gagal karena jaringan disimpan berurutan dan dikirim
// satu per satu saat server terjangkau. Satu berkas JSON di userData berisi
// pesan antrean, transaksi lokal (struk untuk riwayat/cetak ulang), dan delta
// stok yang belum terkirim.

const nodeFs = require('node:fs')
const crypto = require('node:crypto')
const { tulisAtomik, bacaDenganPemulihan } = require('../lib/berkas-atomik')

const STATUS = Object.freeze({ MENUNGGU: 'MENUNGGU', MENGIRIM: 'MENGIRIM', TERKIRIM: 'TERKIRIM', TINJAU: 'TINJAU' })

const LABEL = Object.freeze({
  CHECKOUT: 'Transaksi',
  PENGELUARAN: 'Pengeluaran',
  STOK_MASUK: 'Stok masuk',
  SESI_BUKA: 'Buka sesi'
})

function labelJenis(jenis) { return LABEL[jenis] || jenis }

function buatClientRef() {
  return crypto.randomUUID()
}

/**
 * `waktu_klien` untuk badan permintaan: ISO-8601 **waktu lokal kasir**, detik
 * penuh, tanpa akhiran zona.
 *
 * Bukan `toISOString()`: itu menghasilkan "…T07:25:40.635Z" (UTC + Z), dan
 * server menyimpannya apa adanya ke kolom MySQL `datetime` → error 1292
 * "Incorrect datetime value" sehingga transaksi ditolak. Waktu lokal juga yang
 * dimaksud kolom itu: kapan kasir menekan Bayar menurut jam tokonya.
 */
function waktuKlien(waktu = new Date()) {
  const dua = (n) => String(n).padStart(2, '0')
  return `${waktu.getFullYear()}-${dua(waktu.getMonth() + 1)}-${dua(waktu.getDate())}`
    + `T${dua(waktu.getHours())}:${dua(waktu.getMinutes())}:${dua(waktu.getSeconds())}`
}

/** Bentuk berkas antrean yang sah (objek dengan `pesan` berupa array). */
function bentukAntreanSah(d) {
  return !!d && typeof d === 'object' && !Array.isArray(d) && Array.isArray(d.pesan)
}

class Antrean {
  /**
   * @param {object} o
   * @param {string|null} o.berkas      null = hanya memori (tes)
   * @param {(galat: {jenis: 'simpan'|'muat', pesan: string, detail?: string, berkasRusak?: string[]}) => void} [o.onGalat]
   *        dipanggil saat berkas gagal ditulis / rusak saat dimuat — ipc.js meneruskan ke
   *        status UI dan laporan diagnostik. Penjualan TIDAK boleh hilang diam-diam.
   */
  constructor({ berkas = null, sekarang = () => Date.now(), fsImpl = nodeFs, onGalat = null } = {}) {
    this.berkas = berkas
    this.sekarang = sekarang
    this.fsImpl = fsImpl
    this.onGalat = onGalat
    this.pesan = []        // urut naik
    this.transaksi = {}    // clientRef → { tokoId, nomorLokal, struk, waktuKlien, ... }
    this.delta = []        // { clientRef, idProduk, delta }
    this._urut = 0
    this.galatSimpan = null    // { pesan, detail, pada } — penulisan terakhir gagal (data hanya di memori)
    this.peringatanMuat = null // { pesan, berkasRusak, pada } — berkas rusak saat aplikasi dibuka
    this._muat()
  }

  _lapor(galat) {
    if (this.onGalat) { try { this.onGalat(galat) } catch { /* abaikan */ } }
  }

  _muat() {
    if (!this.berkas) return
    const hasil = bacaDenganPemulihan(this.berkas, bentukAntreanSah, {
      fsImpl: this.fsImpl,
      sekarang: () => new Date(this.sekarang()),
      nama: 'Berkas antrean offline'
    })
    if (hasil.data) {
      const d = hasil.data
      this.pesan = d.pesan
      this.transaksi = d.transaksi && typeof d.transaksi === 'object' ? d.transaksi : {}
      this.delta = Array.isArray(d.delta) ? d.delta : []
      this._urut = this.pesan.reduce((m, p) => Math.max(m, p.urut || 0), 0)
      // Proses mati saat MENGIRIM → kembalikan ke MENUNGGU agar dicoba lagi.
      for (const p of this.pesan) if (p.status === STATUS.MENGIRIM) p.status = STATUS.MENUNGGU
    }
    if (hasil.peringatan) {
      this.peringatanMuat = { pesan: hasil.peringatan, berkasRusak: hasil.berkasRusak, pada: this.sekarang() }
      this._lapor({ jenis: 'muat', pesan: hasil.peringatan, berkasRusak: hasil.berkasRusak })
    }
  }

  /**
   * Tulis antrean ke disk (atomik + generasi .bak). Mengembalikan true bila tersimpan.
   * Gagal → `galatSimpan` diisi & dilaporkan; data tetap di memori dan penulisan
   * berikutnya mencoba lagi.
   */
  simpan() {
    if (!this.berkas) return true
    try {
      tulisAtomik(this.berkas, JSON.stringify({ pesan: this.pesan, transaksi: this.transaksi, delta: this.delta }), { fsImpl: this.fsImpl })
      if (this.galatSimpan) this.galatSimpan = null
      return true
    } catch (err) {
      const pertama = !this.galatSimpan
      this.galatSimpan = {
        pesan: 'Antrean offline gagal disimpan ke komputer ini. Jangan tutup aplikasi sampai data terkirim atau masalah disk diatasi.',
        detail: err && err.message ? String(err.message) : String(err),
        pada: this.sekarang()
      }
      if (pertama) this._lapor({ jenis: 'simpan', pesan: this.galatSimpan.pesan, detail: this.galatSimpan.detail })
      return false
    }
  }

  /** Tandai peringatan muat sudah dilihat pengguna. */
  abaikanPeringatanMuat() { this.peringatanMuat = null }

  /** Antrekan pesan (+ transaksi lokal & delta stok) dalam satu penulisan. */
  antrekan({ clientRef, jenis, tokoId = null, path: jalur, body, status = STATUS.MENUNGGU, galat = null }, { transaksi = null, deltaStok = {} } = {}) {
    const p = {
      urut: ++this._urut,
      clientRef,
      jenis,
      tokoId,
      path: jalur,
      body,
      dibuat: this.sekarang(),
      status,
      percobaan: 0,
      cobaLagiSetelah: null,
      galat,
      hasil: null
    }
    this.pesan.push(p)
    if (transaksi) this.transaksi[clientRef] = { ...transaksi, clientRef }
    for (const [idProduk, d] of Object.entries(deltaStok)) {
      if (Number(d)) this.delta.push({ clientRef, idProduk, delta: Number(d) })
    }
    this.simpan()
    return p
  }

  semua() { return this.pesan.slice().sort((a, b) => a.urut - b.urut) }
  aktif() { return this.semua().filter((p) => p.status !== STATUS.TERKIRIM) }
  cari(clientRef) { return this.pesan.find((p) => p.clientRef === clientRef) || null }

  perbarui(clientRef, patch) {
    const p = this.cari(clientRef)
    if (!p) return null
    Object.assign(p, patch)
    this.simpan()
    return p
  }

  /** Terkirim: simpan hasil, buang transaksi lokal & delta stoknya. */
  selesai(clientRef, hasil) {
    const p = this.cari(clientRef)
    if (!p) return
    p.status = STATUS.TERKIRIM
    p.hasil = hasil || null
    p.cobaLagiSetelah = null
    p.galat = null
    delete this.transaksi[clientRef]
    this.delta = this.delta.filter((d) => d.clientRef !== clientRef)
    // Riwayat terkirim dipangkas agar berkas tak membengkak.
    const terkirim = this.pesan.filter((x) => x.status === STATUS.TERKIRIM)
    if (terkirim.length > 100) {
      const buang = new Set(terkirim.slice(0, terkirim.length - 100).map((x) => x.clientRef))
      this.pesan = this.pesan.filter((x) => !buang.has(x.clientRef))
    }
    this.simpan()
  }

  /** Batalkan (pengguna): buang pesan, transaksi lokal, delta stok. */
  batalkan(clientRef) {
    this.pesan = this.pesan.filter((p) => p.clientRef !== clientRef)
    delete this.transaksi[clientRef]
    this.delta = this.delta.filter((d) => d.clientRef !== clientRef)
    this.simpan()
  }

  ringkas(tokoId = null) {
    const cocok = (p) => tokoId == null || p.tokoId == null || p.tokoId === tokoId
    let menunggu = 0; let tinjau = 0
    for (const p of this.pesan) {
      if (!cocok(p)) continue
      if (p.status === STATUS.MENUNGGU || p.status === STATUS.MENGIRIM) menunggu++
      else if (p.status === STATUS.TINJAU) tinjau++
    }
    return { menunggu, tinjau, total: menunggu + tinjau }
  }

  /**
   * Jumlah baris yang masih menunggu setelah ≥ `ambang` galat server beruntun (muatan
   * yang terus ditolak 5xx) — panel Sinkronisasi menampilkan peringatan menetap.
   */
  macet(tokoId = null, ambang = Infinity) {
    const cocok = (p) => tokoId == null || p.tokoId == null || p.tokoId === tokoId
    return this.pesan.filter((p) => cocok(p) && (p.status === STATUS.MENUNGGU || p.status === STATUS.MENGIRIM) && (p.galatServer || 0) >= ambang).length
  }

  /** Masalah penyimpanan untuk pita status UI (null = sehat). */
  masalahPenyimpanan() {
    if (this.galatSimpan) return { jenis: 'simpan', pesan: this.galatSimpan.pesan, pada: this.galatSimpan.pada }
    if (this.peringatanMuat) return { jenis: 'muat', pesan: this.peringatanMuat.pesan, pada: this.peringatanMuat.pada, berkasRusak: this.peringatanMuat.berkasRusak }
    return null
  }

  transaksiTertunda(tokoId = null) {
    return Object.values(this.transaksi)
      .filter((t) => tokoId == null || t.tokoId == null || t.tokoId === tokoId)
      .sort((a, b) => String(b.waktuKlien).localeCompare(String(a.waktuKlien)))
  }

  transaksiLokal(clientRef) { return this.transaksi[clientRef] || null }

  deltaStokTertunda(tokoId = null) {
    const out = {}
    for (const d of this.delta) {
      const t = this.transaksi[d.clientRef]
      if (tokoId != null && t && t.tokoId && t.tokoId !== tokoId) continue
      out[d.idProduk] = (out[d.idProduk] || 0) + d.delta
    }
    return out
  }
}

/**
 * Rapikan `waktu_klien` pada badan yang sudah tersimpan di antrean sebelum
 * dikirim. Baris yang diantrekan versi lama membawa "…Z" (UTC) yang ditolak
 * MySQL; tanpa ini transaksi lama akan terus gagal walau aplikasi sudah baru.
 */
function badanWaktuRapi(body) {
  const nilai = body && body.waktu_klien
  if (typeof nilai !== 'string' || !nilai) return body
  const t = new Date(nilai)
  if (Number.isNaN(t.getTime())) return body
  const rapi = waktuKlien(t)
  return rapi === nilai ? body : { ...body, waktu_klien: rapi }
}

module.exports = { Antrean, STATUS, labelJenis, buatClientRef, waktuKlien, badanWaktuRapi, bentukAntreanSah }

File: mobile-flutter/lib/core/offline/antrean.dart
import 'dart:convert';

import 'package:drift/drift.dart';

import 'salinan_db.dart';

/// Status satu pesan antrean.
enum StatusAntrean { menunggu, mengirim, terkirim, tinjau }

StatusAntrean statusDari(String s) => switch (s.toUpperCase()) {
  'MENGIRIM' => StatusAntrean.mengirim,
  'TERKIRIM' => StatusAntrean.terkirim,
  'TINJAU' => StatusAntrean.tinjau,
  _ => StatusAntrean.menunggu,
};

String statusKe(StatusAntrean s) => s.name.toUpperCase();

/// Satu permintaan tulis yang menunggu dikirim (baris `outbox`).
class PesanAntrean {
  const PesanAntrean({
    required this.urut,
    required this.clientRef,
    required this.jenis,
    required this.path,
    required this.body,
    required this.dibuat,
    this.tokoId,
    this.pemilik,
    this.percobaan = 0,
    this.galatServer = 0,
    this.cobaLagiSetelah,
    this.status = StatusAntrean.menunggu,
    this.galatTerakhir,
    this.hasil,
  });

  final int urut;
  final String clientRef;

  /// CHECKOUT | PENGELUARAN | STOK_MASUK
  final String jenis;
  final String? tokoId;

  /// Id akun yang membuat baris ini. Antrean hanya dikirim dengan token akun
  /// yang sama: setelah sesi berakhir (401) lalu akun LAIN masuk di perangkat
  /// ini, penjualan akun sebelumnya tidak boleh tercatat atas nama akun baru.
  /// null = baris dari versi aplikasi sebelum penanda ini ada.
  final String? pemilik;
  final String path;
  final Map<String, dynamic> body;
  final DateTime dibuat;
  final int percobaan;

  /// Jumlah jawaban GANGGUAN server (5xx/408/429) — lihat [BatasAntrean].
  final int galatServer;
  final DateTime? cobaLagiSetelah;
  final StatusAntrean status;
  final String? galatTerakhir;
  final Map<String, dynamic>? hasil;

  bool get perluPerhatian => status == StatusAntrean.tinjau;
  bool get belumTerkirim =>
      status == StatusAntrean.menunggu || status == StatusAntrean.mengirim;

  String get label => switch (jenis) {
    'CHECKOUT' => 'Transaksi',
    'PENGELUARAN' => 'Pengeluaran',
    'STOK_MASUK' => 'Stok masuk',
    'OPNAME' => 'Opname stok',
    'SESI_BUKA' => 'Buka sesi',
    'BILL_BUKA' => 'Buka bon',
    'BILL_RONDE' => 'Pesanan bon',
    'BILL_BAYAR' => 'Bayar bon',
    _ => jenis,
  };
}

/// Transaksi lokal (belum tersinkron) untuk riwayat & cetak ulang.
class TransaksiTertunda {
  const TransaksiTertunda({
    required this.clientRef,
    required this.nomorLokal,
    required this.tipePembayaran,
    required this.grandTotal,
    required this.dibayar,
    required this.waktuKlien,
    required this.strukJson,
    this.tokoId,
    this.nomorServer,
  });

  final String clientRef;
  final String? tokoId;
  final String nomorLokal;
  final String? nomorServer;
  final String tipePembayaran;
  final double grandTotal;
  final double dibayar;
  final DateTime waktuKlien;
  final String strukJson;
}

/// Kebijakan antrean dari server (`GET /config`, master platform_config):
/// setelah [maksPercobaanGalatServer] jawaban gangguan server, atau bila baris
/// sudah berumur [maksUmurJam] jam dan server masih gangguan, baris pindah ke
/// TINJAU dengan pesan server terakhir. null = server tidak menetapkan →
/// aplikasi TIDAK pernah memindahkannya sendiri (tanpa angka karangan).
class BatasAntrean {
  const BatasAntrean({this.maksPercobaanGalatServer, this.maksUmurJam});

  final int? maksPercobaanGalatServer;
  final int? maksUmurJam;

  static const tanpaBatas = BatasAntrean();

  bool get ditetapkan => maksPercobaanGalatServer != null || maksUmurJam != null;

  static int? _positif(Object? v) {
    final n = v is num ? v.toInt() : int.tryParse('${v ?? ''}'.trim());
    return n != null && n > 0 ? n : null;
  }

  /// Dari `data` jawaban `/config`.
  factory BatasAntrean.dariConfig(Map<String, dynamic>? data) => BatasAntrean(
    maksPercobaanGalatServer: _positif(data?['antrean_maks_percobaan_galat_server']),
    maksUmurJam: _positif(data?['antrean_maks_umur_jam']),
  );

  /// true bila baris yang BARU SAJA gagal karena gangguan server sudah
  /// melewati batas.
  bool terlampaui({required int galatServer, required DateTime dibuat, required DateTime sekarang}) {
    final maks = maksPercobaanGalatServer;
    if (maks != null && galatServer >= maks) return true;
    final umur = maksUmurJam;
    if (umur != null && sekarang.difference(dibuat) >= Duration(hours: umur)) return true;
    return false;
  }
}

/// Ringkasan untuk pita status & layar Sinkronisasi.
class RingkasAntrean {
  const RingkasAntrean({this.menunggu = 0, this.tinjau = 0, this.milikLain = 0});
  final int menunggu;
  final int tinjau;

  /// Baris belum terkirim milik akun lain (menunggu akun itu masuk lagi).
  final int milikLain;
  int get total => menunggu + tinjau;
}

/// true bila [p] milik akun [pemilik] (null = semua dianggap milik sendiri,
/// dipakai uji & isolate latar sebelum akun pernah tercatat).
bool milikAkun(PesanAntrean p, String? pemilik) =>
    pemilik == null || p.pemilik == pemilik;

/// Ringkasan dari sudut pandang akun [pemilik]: baris akun lain dihitung
/// terpisah sebagai [RingkasAntrean.milikLain].
RingkasAntrean ringkasUntuk(Iterable<PesanAntrean> semua, String? pemilik) {
  var menunggu = 0, tinjau = 0, lain = 0;
  for (final p in semua) {
    if (p.status == StatusAntrean.terkirim) continue;
    if (!milikAkun(p, pemilik)) {
      lain++;
    } else if (p.perluPerhatian) {
      tinjau++;
    } else {
      menunggu++;
    }
  }
  return RingkasAntrean(menunggu: menunggu, tinjau: tinjau, milikLain: lain);
}

/// Penyimpanan antrean — antarmuka kecil (memori di test, SQLite di perangkat).
abstract interface class AntreanStore {
  /// Antrekan satu pesan beserta transaksi lokal & delta stok dalam SATU
  /// transaksi penyimpanan: tidak mungkin ada struk tanpa antrean.
  Future<void> antrekan(
    PesanAntrean pesan, {
    TransaksiTertunda? transaksi,
    Map<String, double> deltaStok = const {},
  });

  Future<List<PesanAntrean>> semua();

  /// Pesan MENUNGGU yang boleh dikirim sekarang, urut FIFO.
  Future<List<PesanAntrean>> siapKirim(DateTime sekarang);

  Future<PesanAntrean?> cari(String clientRef);

  Future<void> perbarui(
    String clientRef, {
    StatusAntrean? status,
    int? percobaan,
    int? galatServer,
    DateTime? cobaLagiSetelah,
    bool hapusCobaLagi = false,
    String? galatTerakhir,
    Map<String, dynamic>? hasil,
  });

  /// Selesai terkirim: catat hasil, hapus transaksi lokal & delta stoknya.
  Future<void> selesai(String clientRef, {Map<String, dynamic>? hasil});

  /// Batalkan (pengguna): hapus pesan, transaksi lokal, dan delta stoknya.
  Future<void> batalkan(String clientRef);

  Future<RingkasAntrean> ringkas();

  /// Tandai baris tanpa pemilik (dibuat versi lama) sebagai milik [pemilik].
  /// Dipanggil saat akun yang sama dengan akun terakhir perangkat masuk.
  Future<int> klaimTanpaPemilik(String pemilik);

  Future<List<TransaksiTertunda>> transaksiTertunda({String? tokoId});

  Future<TransaksiTertunda?> transaksiLokal(String clientRef);

  /// Delta stok tertunda per id produk (dijumlahkan).
  Future<Map<String, double>> deltaStokTertunda({String? tokoId});
}

// ---------------------------------------------------------------- memori

class AntreanMemori implements AntreanStore {
  final List<PesanAntrean> _pesan = [];
  final Map<String, TransaksiTertunda> _trx = {};
  final List<(String clientRef, String idProduk, double delta)> _delta = [];
  int _urut = 0;

  @override
  Future<void> antrekan(
    PesanAntrean pesan, {
    TransaksiTertunda? transaksi,
    Map<String, double> deltaStok = const {},
  }) async {
    _pesan.add(_salin(pesan, urut: ++_urut));
    if (transaksi != null) _trx[pesan.clientRef] = transaksi;
    deltaStok.forEach((id, d) => _delta.add((pesan.clientRef, id, d)));
  }

  PesanAntrean _salin(
    PesanAntrean p, {
    int? urut,
    String? pemilik,
    StatusAntrean? status,
    int? percobaan,
    int? galatServer,
    DateTime? cobaLagiSetelah,
    bool hapusCobaLagi = false,
    String? galatTerakhir,
    Map<String, dynamic>? hasil,
  }) => PesanAntrean(
    urut: urut ?? p.urut,
    clientRef: p.clientRef,
    jenis: p.jenis,
    path: p.path,
    body: p.body,
    dibuat: p.dibuat,
    tokoId: p.tokoId,
    pemilik: pemilik ?? p.pemilik,
    percobaan: percobaan ?? p.percobaan,
    galatServer: galatServer ?? p.galatServer,
    cobaLagiSetelah: hapusCobaLagi ? null : (cobaLagiSetelah ?? p.cobaLagiSetelah),
    status: status ?? p.status,
    galatTerakhir: galatTerakhir ?? p.galatTerakhir,
    hasil: hasil ?? p.hasil,
  );

  @override
  Future<List<PesanAntrean>> semua() async =>
      List.unmodifiable(_pesan..sort((a, b) => a.urut.compareTo(b.urut)));

  @override
  Future<List<PesanAntrean>> siapKirim(DateTime sekarang) async => [
    for (final p in await semua())
      if (p.status == StatusAntrean.menunggu &&
          (p.cobaLagiSetelah == null || !p.cobaLagiSetelah!.isAfter(sekarang)))
        p,
  ];

  @override
  Future<PesanAntrean?> cari(String clientRef) async {
    for (final p in _pesan) {
      if (p.clientRef == clientRef) return p;
    }
    return null;
  }

  @override
  Future<void> perbarui(
    String clientRef, {
    StatusAntrean? status,
    int? percobaan,
    int? galatServer,
    DateTime? cobaLagiSetelah,
    bool hapusCobaLagi = false,
    String? galatTerakhir,
    Map<String, dynamic>? hasil,
  }) async {
    final i = _pesan.indexWhere((p) => p.clientRef == clientRef);
    if (i < 0) return;
    _pesan[i] = _salin(
      _pesan[i],
      status: status,
      percobaan: percobaan,
      galatServer: galatServer,
      cobaLagiSetelah: cobaLagiSetelah,
      hapusCobaLagi: hapusCobaLagi,
      galatTerakhir: galatTerakhir,
      hasil: hasil,
    );
  }

  @override
  Future<void> selesai(String clientRef, {Map<String, dynamic>? hasil}) async {
    await perbarui(clientRef, status: StatusAntrean.terkirim, hasil: hasil, hapusCobaLagi: true);
    _trx.remove(clientRef);
    _delta.removeWhere((d) => d.$1 == clientRef);
  }

  @override
  Future<void> batalkan(String clientRef) async {
    _pesan.removeWhere((p) => p.clientRef == clientRef);
    _trx.remove(clientRef);
    _delta.removeWhere((d) => d.$1 == clientRef);
  }

  @override
  Future<RingkasAntrean> ringkas() async => RingkasAntrean(
    menunggu: _pesan.where((p) => p.belumTerkirim).length,
    tinjau: _pesan.where((p) => p.perluPerhatian).length,
  );

  @override
  Future<int> klaimTanpaPemilik(String pemilik) async {
    var n = 0;
    for (var i = 0; i < _pesan.length; i++) {
      if (_pesan[i].pemilik == null) {
        _pesan[i] = _salin(_pesan[i], pemilik: pemilik);
        n++;
      }
    }
    return n;
  }

  @override
  Future<List<TransaksiTertunda>> transaksiTertunda({String? tokoId}) async => [
    for (final t in _trx.values)
      if (tokoId == null || t.tokoId == tokoId) t,
  ]..sort((a, b) => b.waktuKlien.compareTo(a.waktuKlien));

  @override
  Future<TransaksiTertunda?> transaksiLokal(String clientRef) async =>
      _trx[clientRef];

  @override
  Future<Map<String, double>> deltaStokTertunda({String? tokoId}) async {
    final refToko = tokoId == null ? null : _refTokoAntrean(tokoId);
    final out = <String, double>{};
    for (final d in _delta) {
      if (refToko != null && !refToko.contains(d.$1)) continue;
      out[d.$2] = (out[d.$2] ?? 0) + d.$3;
    }
    return out;
  }

  Set<String> _refTokoAntrean(String tokoId) => {
    for (final p in _pesan)
      if (p.tokoId == tokoId || p.tokoId == null) p.clientRef,
  };
}

// ----------------------------------------------------------------- drift

class AntreanDriftStore implements AntreanStore {
  AntreanDriftStore(this._db);
  final SalinanDb _db;

  PesanAntrean _dari(BarisOutbox b) => PesanAntrean(
    urut: b.urut,
    clientRef: b.clientRef,
    jenis: b.jenis,
    tokoId: b.tokoId,
    pemilik: b.pemilik,
    path: b.path,
    body: Map<String, dynamic>.from(jsonDecode(b.bodyJson) as Map),
    dibuat: b.dibuat,
    percobaan: b.percobaan,
    galatServer: b.galatServer,
    cobaLagiSetelah: b.cobaLagiSetelah,
    status: statusDari(b.status),
    galatTerakhir: b.galatTerakhir,
    hasil: b.hasilJson == null
        ? null
        : Map<String, dynamic>.from(jsonDecode(b.hasilJson!) as Map),
  );

  TransaksiTertunda _trxDari(BarisTransaksiLokal b) => TransaksiTertunda(
    clientRef: b.clientRef,
    tokoId: b.tokoId,
    nomorLokal: b.nomorLokal,
    nomorServer: b.nomorServer,
    tipePembayaran: b.tipePembayaran,
    grandTotal: b.grandTotal,
    dibayar: b.dibayar,
    waktuKlien: b.waktuKlien,
    strukJson: b.strukJson,
  );

  @override
  Future<void> antrekan(
    PesanAntrean pesan, {
    TransaksiTertunda? transaksi,
    Map<String, double> deltaStok = const {},
  }) => _db.transaction(() async {
    await _db
        .into(_db.outbox)
        .insert(
          OutboxCompanion.insert(
            clientRef: pesan.clientRef,
            jenis: pesan.jenis,
            tokoId: Value(pesan.tokoId),
            pemilik: Value(pesan.pemilik),
            path: pesan.path,
            bodyJson: jsonEncode(pesan.body),
            status: Value(statusKe(pesan.status)),
            dibuat: pesan.dibuat,
          ),
        );
    if (transaksi != null) {
      await _db
          .into(_db.transaksiLokal)
          .insert(
            TransaksiLokalCompanion.insert(
              clientRef: transaksi.clientRef,
              tokoId: Value(transaksi.tokoId),
              nomorLokal: transaksi.nomorLokal,
              tipePembayaran: transaksi.tipePembayaran,
              grandTotal: transaksi.grandTotal,
              dibayar: transaksi.dibayar,
              waktuKlien: transaksi.waktuKlien,
              strukJson: transaksi.strukJson,
            ),
          );
    }
    for (final e in deltaStok.entries) {
      await _db
          .into(_db.stokDelta)
          .insert(
            StokDeltaCompanion.insert(
              clientRef: pesan.clientRef,
              idProduk: e.key,
              delta: e.value,
            ),
          );
    }
  });

  @override
  Future<List<PesanAntrean>> semua() async {
    final rows = await (_db.select(_db.outbox)..orderBy([(t) => OrderingTerm.asc(t.urut)])).get();
    return rows.map(_dari).toList();
  }

  @override
  Future<List<PesanAntrean>> siapKirim(DateTime sekarang) async => [
    for (final p in await semua())
      if (p.status == StatusAntrean.menunggu &&
          (p.cobaLagiSetelah == null || !p.cobaLagiSetelah!.isAfter(sekarang)))
        p,
  ];

  @override
  Future<PesanAntrean?> cari(String clientRef) async {
    final b = await (_db.select(_db.outbox)..where((t) => t.clientRef.equals(clientRef))).getSingleOrNull();
    return b == null ? null : _dari(b);
  }

  @override
  Future<void> perbarui(
    String clientRef, {
    StatusAntrean? status,
    int? percobaan,
    int? galatServer,
    DateTime? cobaLagiSetelah,
    bool hapusCobaLagi = false,
    String? galatTerakhir,
    Map<String, dynamic>? hasil,
  }) async {
    await (_db.update(_db.outbox)..where((t) => t.clientRef.equals(clientRef))).write(
      OutboxCompanion(
        status: status == null ? const Value.absent() : Value(statusKe(status)),
        percobaan: percobaan == null ? const Value.absent() : Value(percobaan),
        galatServer: galatServer == null ? const Value.absent() : Value(galatServer),
        cobaLagiSetelah: hapusCobaLagi
            ? const Value(null)
            : (cobaLagiSetelah == null ? const Value.absent() : Value(cobaLagiSetelah)),
        galatTerakhir: galatTerakhir == null ? const Value.absent() : Value(galatTerakhir),
        hasilJson: hasil == null ? const Value.absent() : Value(jsonEncode(hasil)),
      ),
    );
  }

  @override
  Future<void> selesai(String clientRef, {Map<String, dynamic>? hasil}) =>
      _db.transaction(() async {
        await perbarui(clientRef, status: StatusAntrean.terkirim, hasil: hasil, hapusCobaLagi: true);
        await (_db.delete(_db.transaksiLokal)..where((t) => t.clientRef.equals(clientRef))).go();
        await (_db.delete(_db.stokDelta)..where((t) => t.clientRef.equals(clientRef))).go();
      });

  @override
  Future<void> batalkan(String clientRef) => _db.transaction(() async {
    await (_db.delete(_db.outbox)..where((t) => t.clientRef.equals(clientRef))).go();
    await (_db.delete(_db.transaksiLokal)..where((t) => t.clientRef.equals(clientRef))).go();
    await (_db.delete(_db.stokDelta)..where((t) => t.clientRef.equals(clientRef))).go();
  });

  @override
  Future<RingkasAntrean> ringkas() async {
    final rows = await semua();
    return RingkasAntrean(
      menunggu: rows.where((p) => p.belumTerkirim).length,
      tinjau: rows.where((p) => p.perluPerhatian).length,
    );
  }

  @override
  Future<int> klaimTanpaPemilik(String pemilik) =>
      (_db.update(_db.outbox)..where((t) => t.pemilik.isNull())).write(
        OutboxCompanion(pemilik: Value(pemilik)),
      );

  @override
  Future<List<TransaksiTertunda>> transaksiTertunda({String? tokoId}) async {
    final q = _db.select(_db.transaksiLokal)..orderBy([(t) => OrderingTerm.desc(t.waktuKlien)]);
    if (tokoId != null) q.where((t) => t.tokoId.equals(tokoId));
    return (await q.get()).map(_trxDari).toList();
  }

  @override
  Future<TransaksiTertunda?> transaksiLokal(String clientRef) async {
    final b = await (_db.select(_db.transaksiLokal)..where((t) => t.clientRef.equals(clientRef))).getSingleOrNull();
    return b == null ? null : _trxDari(b);
  }

  @override
  Future<Map<String, double>> deltaStokTertunda({String? tokoId}) async {
    final rows = await _db.select(_db.stokDelta).get();
    final out = <String, double>{};
    Set<String>? refToko;
    if (tokoId != null) {
      // Toko pemilik delta diambil dari OUTBOX, bukan dari transaksi_lokal:
      // tabel itu hanya berisi checkout kasir, sehingga opname, stok masuk, dan
      // pelunasan bon (yang juga menitipkan delta) dulu selalu tersaring habis
      // dan stok di katalog tidak pernah ikut berubah saat offline.
      final q = _db.select(_db.outbox)
        ..where((t) => t.tokoId.equals(tokoId) | t.tokoId.isNull());
      refToko = (await q.get()).map((p) => p.clientRef).toSet();
    }
    for (final r in rows) {
      if (refToko != null && !refToko.contains(r.clientRef)) continue;
      out[r.idProduk] = (out[r.idProduk] ?? 0) + r.delta;
    }
    return out;
  }
}
